### PR TITLE
Update Table Detail Columns

### PR DESCRIPTION
**Pull request by danwom**
### Summary of Changes

Redesign of Table Detail Columns

### Tests

_What tests did you add or modify and why? If no tests were added or modified, explain why. Remove this line_

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes, including screenshots of any UI changes. 
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public python functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes all tests documented in the [developer guide](https://github.com/lyft/amundsenfrontendlibrary/blob/master/docs/developer_guide.md#testing)

### CODE REVIEW

File: amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import * as React from 'react';
```
Comment (ttannis) [inline amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: This component is missing the support for notifications that existed in `DetailListItem`, I think it goes in that "actions" section?

Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: This particular branch hasn't been merged with notifications/search_v2 yet. I was going to add those things in after merge. Will fix variable reference as well.

Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: I added the `Request Column Description` dropdown, but it looks a little different from mocks. Can touch up here, or change on master first and then merge down. The former being less work.

File: amundsen_application/static/js/components/TableDetail/ColumnList/styles.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.column-list {
+  margin: 0;
+
+  // TODO - Revisit hover styles when merged with search branch
```
Comment (danwom) [inline amundsen_application/static/js/components/TableDetail/ColumnList/styles.scss]: Will revisit now.

**Comment by danwom** [comment]
@ttannis This should be ready for final review.

File: amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx (hunk 1 above)
Comment (ttannis) [reply amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: I don't mind either option.

File: amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+              </div>
+              {
+                !this.state.isExpanded &&
+                <div className="body-secondary-3 truncated">
+                  { metadata.description }
+                </div>
+              }
+            </div>
+            <div className="resource-type">
+              { metadata.type ? metadata.type.toLowerCase() : 'null' }
```
Comment (ttannis) [inline amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: nit: Are there tests for the correct rendering of `name`, `description` and `type`?

File: amundsen_application/static/js/components/TableDetail/ColumnStats/tests/index.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    };
+    const wrapper = shallow<ColumnStats>(<ColumnStats {...props} />);
+    return { props, wrapper };
+  };
+
+  const { wrapper, props } = setup();
+  const instance = wrapper.instance();
+
+  describe('formatDate', () => {
+    it('formats a date in the correct format', () => {
```
Comment (ttannis) [inline amundsen_application/static/js/components/TableDetail/ColumnStats/tests/index.spec.tsx]: nit: Why two similar tests? Does one test some weird edge case?

File: amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    let infoText = 'Stats reflect data collected';
+    if (startDate && endDate) {
+      if (startDate === endDate) {
+        infoText = `${infoText} on ${startDate} only. (daily partition)`;
+      } else {
+        infoText = `${infoText} between ${startDate} and ${endDate}.`;
+      }
+    } else {
+      infoText = `${infoText} over a recent period of time.`;
```
Comment (ttannis) [inline amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx]: We can delete this case if you want. It protects against a case that shouldn't happen -- not having `startEpoch` or `endEpoch`. My thinking here is that maybe we don't want to hide this issue in the UI so that users will actually recognize it as a bug if it ever happens.

File: amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  render = () => {
+    const { stats } = this.props;
+    if (stats.length === 0) {
+      return null;
+    }
+    const startEpoch = Math.min(...stats.map(s => s.start_epoch));
+    const endEpoch = Math.max(...stats.map(s => s.end_epoch));
+
+    return (
```
Comment (ttannis) [inline amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx]: nit: some decoupling will help to write better tests here. We can also write tests that better verify what goes on in those `map` statements. Exact syntax might need tweaking, essentially raising for the future that we can assert that a thing happened for each thing, in exactly the way that it was supposed to.  
```
props.stats.forEach((state, index ) => {
    expect(renderColumnStatSpy).toHaveBeenCalledWith(stat);
});
```
If the logic we are testing has conditional logic, then that logic can also be separated out  in the code and mocked in the test.

File: amundsen_application/static/js/components/TableDetail/ColumnStats/styles.scss [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      }
+
+      .column-stat-row {
+        display: flex;
+        border-top: 1px solid $divider;
+        line-height: 32px;
+
+        &:nth-child(odd) {
+          // TODO - #FCFCFF is used in mocks but is not part of our color palette. Follow up with design about random colors
+          background-color: #FCFCFF;
```
[the remark this answers is not in the document]
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/styles.scss]: I didn't even realize $gray0 was a thing. Maybe `$body-bg-secondary` should be `$gray0` rather than `$gray5`

File: amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            </div>
+            <div className="badges">
+              {/* Placeholder */}
+            </div>
+            <div className="actions">
+              {
+                notificationsEnabled() &&
+                <Dropdown id={`detail-list-item-dropdown:${this.props.index}`}
```
[the remark this answers is not in the document]
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnListItem/index.tsx]: I'll put a TODO here.

File: amundsen_application/static/js/components/TableDetail/ColumnListItem/tests/index.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      expect(setStateSpy).toHaveBeenCalledWith({ isExpanded: !isExpanded });
+      instance.toggleExpand(null);
+      expect(setStateSpy).toHaveBeenCalledWith({ isExpanded });
+    });
+  });
+
+
+  describe('openRequest', () => {
+    it('calls openRequestDescriptionDialog', () => {
+      expect(props.openRequestDescriptionDialog).not.toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnListItem/tests/index.spec.tsx]: Added a new spy here.

File: amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx (hunk 5 above)
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx]: Should we show an error message, an empty string, or actually throw an error?

File: amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx (hunk 6 above)
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx]: Added a TODO to move map functions, also added test case for handling each stat.

File: amundsen_application/static/js/components/TableDetail/ColumnStats/tests/index.spec.tsx (hunk 4 above)
Comment (danwom) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/tests/index.spec.tsx]: I can delete one case. I just get the feeling that date formats are easily broken so they need additional testing :)

**Comment by danwom** [comment]
@ttannis I did some cleanup but it appears the the approval still stands. Maybe the rules are set differently here vs private branches?

File: amundsen_application/static/js/components/TableDetail/ColumnStats/styles.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (ttannis) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/styles.scss]: On board with whatever you think looks the most cohesive.

File: amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx (hunk 5 above)
Comment (ttannis) [reply amundsen_application/static/js/components/TableDetail/ColumnStats/index.tsx]: Unsure. 🤔I think about how this is probably a bigger issue of validating -- somewhere earlier in this process -- that the data we receive from API calls match what the application expects. I don't think it should be the responsibility of the component to account for things that violate the TypeScript definition, rather some earlier step in ducks or the Flask layer. We can table this for the future.

File: amundsen_application/static/js/components/TableDetail/ColumnList/index.tsx [hunk 10]
```diff
@@ -1,34 +1,37 @@
...
-interface DetailListProps {
+import "./styles.scss";
+
+
+interface ColumnListProps {
   columns?: TableColumn[];
 }
 
-const DetailList: React.SFC<DetailListProps> = ({ columns }) => {
+const ColumnList: React.SFC<ColumnListProps> = ({ columns }) => {
```
[the remark this answers is not in the document]
Comment (ttannis) [reply amundsen_application/static/js/components/TableDetail/ColumnList/index.tsx]: > @ttannis I did some cleanup but it appears the the approval still stands. Maybe the rules are set differently here vs private branches?

Ah maybe, assuming you meant "protected branches" like master. If it bugs me enough I'll check the branch protections at some point. General sentiment is that approvals should still stand -- sure bad actors could upload more unapproved after and approval but the repo is protected by the fact that we're still the only ones who can merge.
